### PR TITLE
Enable 8-week release cadence

### PR DESCRIPTION
**Pull request by guineveresaenger**
Prior to this change, all provider releases via this tool only occurred on upstream upgrades. This meant that security and codegen/bridge updates would merge to the default branch, but wait on an upstream release to apply to the provider, potentially resulting in a provider using an ancient version of the bridge.

This pull request sets a maximum ~8-week release cadence for all providers.

During a bridge update, we check if we have released the provider in question within the last 8 weeks.
If we have not, and the upgrade is a non-upstream upgrade, we will add a `needs-release/patch` label to the bridge update pull request.

Doing this will result in more frequent upgrades, ensuring our providers use a recent bridge and pulumi version.

One detail to call out:
- We may choose to set the release cadence to 7 weeks, to ensure we "catch" the update at minimum every 8th week; this PR sets it to 8 weeks exactly.
- We can also count actual bridge releases if we do not want to use time comparisons.

Fixes https://github.com/pulumi/ci-mgmt/issues/726.

### CODE REVIEW

File: upgrade/steps.go [hunk 1]
```diff
@@ -315,6 +349,12 @@ var InformGitHub = stepv2.Func70E("Inform Github", func(
 			}
 		}
 
+		// On non-upstream upgrades, we will create a patch release label
+		// if the provider hasn't been released in 8 weeks.
+		if c.MaintenancePatch && !c.UpgradeProviderVersion {
```
Comment (iwahbe) [inline upgrade/steps.go:354]: Optional nit: I would put these in a `switch` statement, since we only want one release label to trigger.

```go
switch {
case  c.UpgradeProviderVersion && len(target.GHIssues) > 0:
...
case c.MaintenancePatch && !c.UpgradeProviderVersion:
...
}

File: upgrade/steps_helpers.go [hunk 2]
```diff
@@ -4,17 +4,16 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"github.com/Masterminds/semver/v3"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	"golang.org/x/mod/modfile"
+	"golang.org/x/mod/module"
```
Comment (iwahbe) [inline upgrade/steps_helpers.go:10]: Can we undo this change?

File: upgrade/steps_test.go [hunk 3]
```diff
@@ -266,3 +266,93 @@ func TestReleaseLabel(t *testing.T) {
...
+			GoPath: "/Users/myuser/go",
+		}
+
+		repo := ProviderRepo{
+			org:  "pulumi",
+			name: "pulumi-cloudinit",
+		}
+
+		maintenanceRelease(context.Wrap(ctx),
+			repo)
```
Comment (iwahbe) [inline upgrade/steps_test.go:310]:
```suggestion
		maintenanceRelease(context.Wrap(ctx), repo)
```

File: upgrade/steps_test.go [hunk 4]
```diff
@@ -266,3 +266,93 @@ func TestReleaseLabel(t *testing.T) {
...
```
Comment (iwahbe) [inline upgrade/steps_test.go:355]:
```suggestion
		maintenanceRelease(context.Wrap(ctx), repo)
```

File: upgrade/steps.go [hunk 5]
```diff
@@ -261,6 +262,39 @@ func UpgradeProviderVersion(
 	return step.Combined("Update TF Provider", steps...)
 }
 
+var maintenanceRelease = stepv2.Func10E("Check if we should release a maintenance patch", func(
+	ctx context.Context,
+	repo ProviderRepo,
+) error {
```
[the remark this answers is not in the document]
Comment (guineveresaenger) [reply upgrade/steps.go:268]: I agree and this is what I did at first; however [the rest of this codebase does not](https://github.com/pulumi/upgrade-provider/blob/bb4318b313e32435d964af6292b4a72506fc1e50/upgrade/steps.go#L1016) so I've followed that pattern here for consistency.

The test works fine; at least insofar as testing `ctx.MaintenancePatch`.

[the remark this answers is not in the document]
Comment (iwahbe) [reply upgrade/steps.go:268]: The rest of the codebase uses `ctx` as an implicit return because it was written against `step` (v1) which did not allow returning values.

We are trying to migrate over to a style of explicit return values.

File: upgrade/steps_test.go [hunk 6]
```diff
@@ -296,3 +296,92 @@ func TestParseUpstreamProviderOrgFromModVersion(t *testing.T) {
...
+		}
+
+		result := maintenanceRelease(context.Wrap(ctx), repo)
+		assert.NotNil(t, result)
+	})
+}
+
+func TestCheckMaintenancePatchExpiredCadence(t *testing.T) {
+
+	simpleReplay(t, jsonMarshal[[]*step.Step](t, `[
```
Comment (iwahbe) [inline upgrade/steps_test.go:346]: This will need a rebase, since this function was renamed.

**Review by iwahbe — APPROVED** [review]
Needs a rebase, but otherwise LGTM
